aten/src/ATen/MetaInit.cpp

```diff
@@ -6,23 +6,305 @@
+    if (tensor.is_meta()) {
+      return tensor;
+    } else {
+      return meta_init_state.getOrCreateSurrogateTensor(tensor);
+    }
+  };
+
+  replaceTensorArguments(op, s, replace_foreign_tensor);
+}
+
+// Checks whether a `TensorOptions` argument can be parsed from the individual
+// arguments of `schema`.
+bool hasTensorOptions(const FunctionSchema& schema) noexcept {
+  std::array<c10::string_view, 4> tensor_opts = {{
+      "dtype", "layout", "device", "pin_memory"
+  }};
+
+  const std::vector<Argument>& args = schema.arguments();
+  if (args.size() < tensor_opts.size()) {
+    return false;
+  }
+
+  // Checks if the arguments starting at `arg_pos` represent a `TensorOptions`.
+  auto are_tensor_opts = [&tensor_opts](auto arg_pos) noexcept {
+    for (const auto& tensor_opt : tensor_opts) {
+      if (tensor_opt != arg_pos->name()) {
+        return false;
+      }
+      ++arg_pos;
+    }
+    return true;
+  };
+
+  for (auto pos = args.begin(); pos <= args.end() - tensor_opts.size(); ++pos) {
+    if (are_tensor_opts(pos)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// TODO: The following function relies on some common heuristics to determine
+// whether an operator is a tensor factory. Although in practice it is mostly
+// reliable, in long term we need a cleaner solution such as operator tagging.
+bool isTensorFactory(const OperatorHandle& op) noexcept {
+  const FunctionSchema& schema = op.schema();
+
+  // For an operator to be considered a factory it must return a tensor.
+  bool has_tensor_output = std::any_of(
+      schema.returns().begin(), schema.returns().end(),
+      [](const Argument& return_arg) {
+        return return_arg.type() == TensorType::get();
+      });
+
+  if (!has_tensor_output) {
+    return false;
+  }
+
+  // We use a simple heuristic that is similar to Autograd's function generator
+  // tool. We check whether the operator has a `TensorOptions` parameter, which
+  // means that it is (very likely) a factory.
+  return hasTensorOptions(schema);
+}
+
+// Replaces the `device` argument in `s` with the meta device if `op` is a
+// tensor factory.
+void replaceDeviceArgument(const OperatorHandle& op, Stack& s) {
+  if (!isTensorFactory(op)) {
+    return;
+  }
+
+  // It is safe to access the index right away skipping the optional check since
+  // we know for sure that a factory has a `device` argument.
+  int device_idx = op.schema().argumentIndexWithName("device").value();
+
+  IValue& device = torch::jit::peek(s, device_idx, op.schema().arguments().size());
+
+  // This condition should always be true, but let's be pedantic.
+  if (device.isDevice()) {
+    device = Device{kMeta};
+  }
+}
+
 constexpr DispatchKeySet after_meta_init_keyset =
     DispatchKeySet{DispatchKeySet::FULL_AFTER, DispatchKey::MetaInit};
 
-void metaInitFallback(const OperatorHandle& op, DispatchKeySet ks, torch::jit::Stack* s) {
+void metaInitFallback(const OperatorHandle& op, DispatchKeySet ks, Stack* s) {
```

Comment:
Could you explain the flow here a bit with a small example in terms of when this logic would be executed?

Does this mean that within `init_meta` context, for tensors that created via mechanisms like `torch.tensor(..)` etc. that cannot be intercepted, we would still allocate memory for them but if those tensors are used as arguments in other ops, we would replace the arguments to those ops with a similar sized meta tensor?

If so, doesn't this break the contract of not allocating memory at all? What if the tensor constructed via an external mechanism was really large and doesn't fit on the target device? It seems like the proper fix should be a way to intercept things like these (especially torch.tensor(...))

Reply:
> Could you explain the flow here a bit with a small example in terms of when this logic would be executed?

This is a typical dispatcher fallback function that is declared with a `TORCH_LIBRARY_IMPL` (see line 74). It is used as a catch-all function if none of your "backend" kernels could handle a dispatch request. In several backends (not sure if that is the right term to use for them) such as Python mode and Tracer it is used in a similar fashion.

> Does this mean that within init_meta context, for tensors that created via mechanisms like torch.tensor(..) etc. that cannot be intercepted, we would still allocate memory for them but if those tensors are used as arguments in other ops, we would replace the arguments to those ops with a similar sized meta tensor?

Correct. This also applies to tensors that were constructed before the meta-init context and then later used inside the context as an operator argument.

> If so, doesn't this break the contract of not allocating memory at all? What if the tensor constructed via an external mechanism was really large and doesn't fit on the target device? It seems like the proper fix should be a way to intercept things like these (especially torch.tensor(...))

There are two cases:
 - If the tensor is constructed by taking ownership of the external data (e.g. `from_numpy`), this does not break the contract. The memory for the external data was already allocated and we are just wrapping it in a tensor. As mentioned above, those tensors will be replaced with surrogates for further operator calls. So we won't modify or unintentionally duplicate the data of such tensors.
 - The second case is problematic and I wrote it in the doc (second PR) as a "failure" case as well. If the external data gets copied (e.g. `torch.tensor()`) then we are breaking the contract. There are ways to prevent this by extending the implementation of [`internal_new_from_data()`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/tensor_new.cpp#L218), but I am not touching it for now since I don't want to extend the scope of this work too much before all stakeholders agree on this approach.